### PR TITLE
OCPBUGS-21814: bump library-go and k8s dependencies to most recent version

### PR DESCRIPTION
**Pull request by cdoern**
bump golang.org/x/net to v0.17.0. I believe this is the only remediation needed by the MCO, at least at the moment.

Also ensure that all http servers have http/2 disabled by default.

### CODE REVIEW

**Comment by rioliu-rh** [comment]
/hold for QE verification
We will run regression test against the cluster built based on this PR

**Comment by rioliu-rh** [comment]
pre-merge testing result
built image with this PR 4.14.0-0.ci.test-2023-10-19-013335-ci-ln-fvvzjmb-latest
setup cluster on AWS and do scale up to add 2 RHEL8 nodes
run automation to execute critical cases w/ and w/o RHEL nodes
test result LGTM
```console
10-19 18:25:10.523  The Case Execution Summary:
10-19 18:25:10.523   PASS OCP-42347 Author:rioliu health check for machine-config-operator [Serial]
10-19 18:25:10.523   PASS OCP-42361 Author:rioliu add chrony systemd config [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-42364 Author:mhanss add selinux kernel argument [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-42365 Author:mhanss add real time kernel argument [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-42367 Author:mhanss add extension to RHCOS [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-42368 Author:mhanss add max pods to the kubelet config [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-42390 Author:rioliu add machine config without ignition version. Block the MCO upgrade rollout if any of the pools are Degraded [Serial]
10-19 18:25:10.523   PASS OCP-42438 Author:mhanss add journald systemd config [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-43048 Author:mhanss create/delete custom machine config pool [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-43064 Author:mhanss create/delete custom machine config pool [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-43310 Author:mhanss add kernel arguments, kernel type and extension to the RHCOS and RHEL [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-45318 Author:rioliu add machine config without ignition version. Block the MCO upgrade rollout if any of the pools are Degraded [Serial]
10-19 18:25:10.523   SKIP OCP-53960 Author:sregidor No failed units in the bootstrap machine
10-19 18:25:10.523   PASS OCP-54085 Author:sregidor Update osImage changing /etc /usr and rpm [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-54159 Author:sregidor Apply a new osImage on a cluster with already installed rpms [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-56131 Author:sregidor Install all extensions [Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-62084 Author:sregidor Certificate rotation in paused pools[Disruptive] [Serial]
10-19 18:25:10.523   PASS OCP-64781 Author:sregidor MCO should be compliant with CIS benchmark rule
10-19 18:25:10.523   PASS OCP-67395 Author:sregidor rotate kubernetes certificate authority. Certificates managed via non-MC path.[Disruptive] [Serial]
10-19 19:09:47.979   PASS OCP-42369 Author:mhanss add container runtime config [Disruptive] [Serial]
```
setup cluster with 4.13.17 build and do upgrade from 4.13.17 to 4.14.0-0.ci.test-2023-10-19-013335-ci-ln-fvvzjmb-latest
upgrade is completed successfully
```console
$ cv version -o yaml | yq -y '.status.history'
- acceptedRisks: 'Target release version="" image="registry.build05.ci.openshift.org/ci-ln-fvvzjmb/release:latest"
    cannot be verified, but continuing anyway because the update was forced: release
    images that are not accessed via digest cannot be verified

    Forced through blocking failures: Multiple precondition checks failed:

    * Precondition "ClusterVersionUpgradeable" failed because of "AdminAckRequired":
    Kubernetes 1.27 and therefore OpenShift 4.14 remove several APIs which require
    admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958395
    for details and instructions.

    * Precondition "EtcdRecentBackup" failed because of "ControllerStarted": RecentBackup:
    The etcd backup controller is starting, and will decide if recent backups are
    available or if a backup is required

    * Precondition "ClusterVersionRecommendedUpdate" failed because of "UnknownUpdate":
    RetrievedUpdates=False (VersionNotFound), so the recommended status of updating
    from 4.13.17 to 4.14.0-0.ci.test-2023-10-19-013335-ci-ln-fvvzjmb-latest is unknown.'
  completionTime: '2023-10-19T11:57:37Z'
  image: registry.build05.ci.openshift.org/ci-ln-fvvzjmb/release:latest
  startedTime: '2023-10-19T10:51:51Z'
  state: Completed
  verified: false
  version: 4.14.0-0.ci.test-2023-10-19-013335-ci-ln-fvvzjmb-latest
- completionTime: '2023-10-19T10:32:12Z'
  image: quay.io/openshift-release-dev/ocp-release@sha256:c1f2fa2170c02869484a4e049132128e216a363634d38abf292eef181e93b692
  startedTime: '2023-10-19T10:15:36Z'
  state: Completed
  verified: false
  version: 4.13.17
```
/unhold
/label qe-approved

**Comment by cdoern** [comment]
our API migration PR took care of the net v0.17.0 bump, this just bumps library-go

File: pkg/controller/common/metrics.go [hunk 1]
```diff
@@ -81,7 +82,7 @@ func StartMetricsListener(addr string, stopCh <-chan struct{}, registerFunc func
 	klog.Infof("Starting metrics listener on %s", addr)
 	mux := http.NewServeMux()
 	mux.Handle("/metrics", promhttp.Handler())
-	s := http.Server{Addr: addr, Handler: mux}
+	s := http.Server{TLSNextProto: make(map[string]func(*http.Server, *tls.Conn, http.Handler)), Addr: addr, Handler: mux}
```
Comment (yuqi-zhang) [inline pkg/controller/common/metrics.go:85]: Based on the conversation in slack, I think only the metrics server will be affected. Were you able to test whether this passes the listed tests for port 9001?

**Comment by rioliu-rh** [comment]
check metric server config and verify whether metrics can be retrieved successfully
```console
$ oc get svc -n openshift-machine-config-operator -ojsonpath='{.items[*].spec.clusterIP}'
172.30.241.164 172.30.8.218 172.30.235.86%
$ oc get svc -n openshift-machine-config-operator -ojsonpath='{.items[*].spec.ports[?(@.name=="metrics")].port}'
9001 9001 9001%

$ oc rsh -n openshift-monitoring prometheus-k8s-0 bash -c 'curl -s -k  -H "Authorization: Bearer `cat /var/run/secrets/kubernetes.io/serviceaccount/token`" https://172.30.8.218:9001/metrics | egrep ^mc | grep -v "#"'
mcd_host_os_and_version{os="RHCOS",version="414.92.202310180038-0"} 1
mcd_kubelet_state 0
mcd_pivot_errors_total 0
mcd_reboots_failed_total 0
mcd_state{reason="",state="Done"} 1.6977758881825037e+09
mcd_update_state{config="MachineConfig: rendered-worker-4eec42896903a91dd704075f19003149",err=""} 1.6977758881915743e+09

$ oc rsh -n openshift-monitoring prometheus-k8s-0 bash -c 'curl -s -k  -H "Authorization: Bearer `cat /var/run/secrets/kubernetes.io/serviceaccount/token`" https://172.30.241.164:9001/metrics | egrep ^mc | grep -v "#"'
mcc_pool_alert{node="ip-10-0-13-243.us-east-2.compute.internal"} 0
mcc_pool_alert{node="ip-10-0-4-73.us-east-2.compute.internal"} 0
mcc_pool_alert{node="ip-10-0-40-233.us-east-2.compute.internal"} 0
mcc_pool_alert{node="ip-10-0-43-21.us-east-2.compute.internal"} 0
mcc_pool_alert{node="ip-10-0-68-113.us-east-2.compute.internal"} 0
mcc_pool_alert{node="ip-10-0-92-155.us-east-2.compute.internal"} 0

$ oc rsh -n openshift-monitoring prometheus-k8s-0 bash -c 'curl -s -k  -H "Authorization: Bearer `cat /var/run/secrets/kubernetes.io/serviceaccount/token`" https://172.30.235.86:9001/metrics | egrep ^mc | grep -v "#"'
mco_degraded_machine_count{pool="master"} 0
mco_degraded_machine_count{pool="worker"} 0
mco_machine_count{pool="master"} 3
mco_machine_count{pool="worker"} 3
mco_state{pool="master",reason="",state="Degraded"} 1.6977758792957501e+09
mco_state{pool="master",reason="",state="Updated"} 1.697779956582294e+09
mco_state{pool="master",reason="",state="Updating"} 1.6977759182945852e+09
mco_state{pool="master",reason="1 nodes are reporting degraded status on sync",state="NodeDegraded"} 1.6977758972941613e+09
mco_state{pool="worker",reason="",state="Degraded"} 1.6977758792957551e+09
mco_state{pool="worker",reason="",state="Updated"} 1.6977799565823002e+09
mco_state{pool="worker",reason="",state="Updating"} 1.6977759112943006e+09
mco_state{pool="worker",reason="1 nodes are reporting degraded status on sync",state="NodeDegraded"} 1.6977758972941658e+09
mco_unavailable_machine_count{pool="master"} 0
mco_unavailable_machine_count{pool="worker"} 0
mco_updated_machine_count{pool="master"} 3
mco_updated_machine_count{pool="worker"} 3
```

**Comment by rioliu-rh** [comment]
Executed alert related regression tests, no issue found
```console
10-20 14:45:14.758  The Case Execution Summary:
10-20 14:45:14.758   PASS OCP-43151 Author:sregidor add node label to service monitor [Serial]
10-20 14:45:14.758   PASS OCP-56706 Author:sregidor Move MCD drain alert into the MCC, revisit error mode[Disruptive] 
[Serial]
10-20 16:06:26.107   PASS OCP-62075 Author:sregidor MCCPoolAlert. Test support for a node pool hierarchy [Disruptive] [Serial]
10-20 16:06:26.107   PASS OCP-63865 Author:sregidor MCDRebootError alert[Disruptive] [Serial]
10-20 16:06:26.107   PASS OCP-63866 Author:sregidor MCDPivotError alert[Disruptive] [Serial]
```

**Comment by sinnykumari** [comment]
I am a bit curious, we bumped k8s deps and library-go but I don't see respective vendor packages being updated. Did you run `make go-deps` or perhaps changes are still in your local repo?

**Comment by cdoern** [comment]
@sinnykumari I noticed that too. I think we just do not use alot or any of the changed library-go code. As for k8s I am not sure but have run make go-deps many times. The library-go changes were ~50 lines and the k8s changes might have been minor as well (for what we use anyway)

**Comment by sinnykumari** [comment]
/hold cancel
LGTM
I think Jerry has been looking into this as well, so will let Jerry do another round of review and see if we want to incorporate any additional changes related to the bug.

**Comment by yuqi-zhang** [comment]
Ignore my previous comment, after some slack discussion we were verifying via the wrong method. We've updated to unencrypted verification the output is as we expect.

/lgtm
/hold

in case we want to revisit testing with the new changes, but can unhold on monday if no further objections

**Comment by rioliu-rh** [comment]
build image with latest code
setup a cluster and provision 2 rhel nodes
run critical cases
```console
The Case Execution Summary:
PASS OCP-42347 Author:rioliu health check for machine-config-operator [Serial]
PASS OCP-42361 Author:rioliu add chrony systemd config [Disruptive] [Serial]
PASS OCP-42364 Author:mhanss add selinux kernel argument [Disruptive] [Serial]
PASS OCP-42365 Author:mhanss add real time kernel argument [Disruptive] [Serial]
PASS OCP-42367 Author:mhanss add extension to RHCOS [Disruptive] [Serial]
PASS OCP-42368 Author:mhanss add max pods to the kubelet config [Disruptive] [Serial]
PASS OCP-42390 Author:rioliu add machine config without ignition version. Block the MCO upgrade rollout if any of the pools are Degraded [Serial]
PASS OCP-42438 Author:mhanss add journald systemd config [Disruptive] [Serial]
PASS OCP-43048 Author:mhanss create/delete custom machine config pool [Disruptive] [Serial]
PASS OCP-43064 Author:mhanss create/delete custom machine config pool [Disruptive] [Serial]
PASS OCP-43310 Author:mhanss add kernel arguments, kernel type and extension to the RHCOS and RHEL [Disruptive] [Serial]
PASS OCP-45318 Author:rioliu add machine config without ignition version. Block the MCO upgrade rollout if any of the pools are Degraded [Serial]
PASS OCP-56131 Author:sregidor Install all extensions [Disruptive] [Serial]
PASS OCP-62084 Author:sregidor Certificate rotation in paused pools[Disruptive] [Serial]
PASS OCP-64781 Author:sregidor MCO should be compliant with CIS benchmark rule
PASS OCP-67395 Author:sregidor rotate kubernetes certificate authority. Certificates managed via non-MC path.
[Disruptive] [Serial]
PASS OCP-54085 Author:sregidor Update osImage changing /etc /usr and rpm [Disruptive] [Serial]
PASS OCP-54159 Author:sregidor Apply a new osImage on a cluster with already installed rpms [Disruptive] [Serial]
```
run alert cases
```
The Case Execution Summary:
PASS OCP-62075 Author:sregidor MCCPoolAlert. Test support for a node pool hierarchy [Disruptive] [Serial]
PASS OCP-63865 Author:sregidor MCDRebootError alert[Disruptive] [Serial]
PASS OCP-63866 Author:sregidor MCDPivotError alert[Disruptive] [Serial]
```

**Comment by rioliu-rh** [comment]
setup cluster with 4.13.18 and upgrade it to the image built based on this PR
upgrade is completed successfully
```console
$ cv version -o yaml | yq -y '.status.history'
- acceptedRisks: 'Target release version="" image="registry.build01.ci.openshift.org/ci-ln-hs6d4ft/release:latest"
    cannot be verified, but continuing anyway because the update was forced: release
    images that are not accessed via digest cannot be verified

    Forced through blocking failures: Multiple precondition checks failed:

    * Precondition "ClusterVersionUpgradeable" failed because of "AdminAckRequired":
    Kubernetes 1.27 and therefore OpenShift 4.14 remove several APIs which require
    admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958395
    for details and instructions.

    * Precondition "EtcdRecentBackup" failed because of "ControllerStarted": RecentBackup:
    The etcd backup controller is starting, and will decide if recent backups are
    available or if a backup is required

    * Precondition "ClusterVersionRecommendedUpdate" failed because of "UnknownUpdate":
    RetrievedUpdates=False (VersionNotFound), so the recommended status of updating
    from 4.13.18 to 4.14.0-0.ci.test-2023-10-23-010215-ci-ln-hs6d4ft-latest is unknown.'
  completionTime: '2023-10-23T13:34:59Z'
  image: registry.build01.ci.openshift.org/ci-ln-hs6d4ft/release:latest
  startedTime: '2023-10-23T12:26:02Z'
  state: Completed
  verified: false
  version: 4.14.0-0.ci.test-2023-10-23-010215-ci-ln-hs6d4ft-latest
- completionTime: '2023-10-23T03:05:13Z'
  image: quay.io/openshift-release-dev/ocp-release@sha256:d0fd9d3ab8690605f816c879d74f4e6d6d9f72982f63a3e0ef3e027ecc512e1c
  startedTime: '2023-10-23T02:45:04Z'
  state: Completed
  verified: false
  version: 4.13.18
```

**Pull request merged**